src/zarr/api/asynchronous.py

```diff
@@ -225,6 +225,11 @@ async def open(
     if path is not None:
         store_path = store_path / path
 
+    if "shape" not in kwargs and mode in {"a", "w", "w-"}:
+        try:
+            return await open_group(store=store_path, zarr_format=zarr_format, mode=mode, **kwargs)
+        except AssertionError:
```

Comment:
So one issue with v3 is that we can't trivially determine if `store_path` corresponds to a group or array by checking if it contains `.zarray` or `.zgroup` like we can with v2, so the `zarr.json` needs to be decoded one way or another. Would it be better to explicitly do this and check the JSON rather than catching an `AssertionError`?

Reply:
I guess either way `open_group` will try and decode `zarr.json`, so opening an array with the current code will result in two attempts at decoding (once in open_group, then again in open_array)? In that case I think it's worth just checking the JSON here. A helper `is_group_or_array(store, path)` function would be nice to use here.

Reply:
How does this look? I moved the code for reading the array metadata into a separate function (`get_array_metadata`) and then used it to check if the path contains an array.